libphal/phal_pdbg.C

```diff
@@ -323,4 +326,121 @@ bool hasControlTransitionedToHost()
+		    locCode);
+		return std::nullopt;
+	}
+
+	// Given location code length should be need to match with minimum
+	// length
+	// to drop expanded format in given location code.
+	if (locCode.length() < EXP_LOCATIN_CODE_MIN_LENGTH) {
+		log(level::ERROR,
+		    "Given location code %s is not meet "
+		    "with minimum length %d",
+		    locCode, EXP_LOCATIN_CODE_MIN_LENGTH);
+		return std::nullopt;
+	}
+
+	// "-" should be there to seggregate all (FC, Node number and SE values)
+	// together.
+	// Note: Each (FC, Node number and SE) value can be seggregate by "."
+	// but, cec device tree just have unexpand format so, just skipping
+	// still first occurrence "-" and replacing with "fcs".
+	auto endPosOfFcs = locCode.find('-', EXP_LOCATIN_CODE_MIN_LENGTH);
+	if (endPosOfFcs == std::string::npos) {
+		log(level::ERROR,
+		    "Given location code %s is not valid i.e could not find "
+		    "dash",
+		    locCode);
+		return std::nullopt;
+	}
+
+	std::string unExpandedLocCode("Ufcs");
+
+	if (locCode.length() > EXP_LOCATIN_CODE_MIN_LENGTH) {
+		unExpandedLocCode.append(
+		    locCode.substr(endPosOfFcs, std::string::npos));
+	}
+
+	return unExpandedLocCode;
+}
+
+int getAttrType(const LocationCode &locationCode, ATTR_TYPE_Type &fruType)
+{
+	ATTR_LOCATION_CODE_Type locCode;
+	// Defining a list of existing frus
+	// Note: We need to update the list if a new fru is added
+	std::set<const char *> frus{"bmc", "proc", "tpm", "dimm"};
+	struct pdbg_target *target;
+
+	for (auto fru : frus) {
+		pdbg_for_each_class_target(fru, target)
+		{
+			if (DT_GET_PROP(ATTR_LOCATION_CODE, target, locCode)) {
+				continue;
+			} else {
+				if (locCode == locationCode) {
+					if (DT_GET_PROP(ATTR_TYPE, target,
+							fruType)) {
+						return openpower::phal::
+						    exception::
+							DEVTREE_ATTR_READ_FAIL;
+					} else {
+						return 0; // success
+					}
+				}
+			}
+		}
+	}
+	return -1;
+}
+
+std::expected<FRUType, openpower::phal::exception::ERR_TYPE>
+    getFRUType(const LocationCode &locationCode) noexcept
+{
+	LocationCode locCode = locationCode;
+	if (!locCode.starts_with("Ufcs-")) {
+
+		auto loc = getUnexpandedLocCode(locationCode);
+		if (loc.has_value()) {
+			locCode = loc.value();
+		} else {
+			log(level::ERROR, "Given location code %s is not valid",
+			    loc);
+			return std::unexpected(openpower::phal::exception::
+						   ATTR_LOC_CODE_NOT_VALID);
+		}
+	}
+	if (locCode.length() < UNEXP_LOCATIN_CODE_MIN_LENGTH) {
+		log(level::ERROR, "Given location code %s is not valid",
+		    locationCode);
+		return std::unexpected(
+		    openpower::phal::exception::ATTR_LOC_CODE_NOT_VALID);
+	}
+	ATTR_TYPE_Type fruType = ENUM_ATTR_TYPE_NA;
+	auto ret = getAttrType(locCode, fruType);
+	if (ret == 0) {
+		return fruType;
+	} else {
+		if (ret == openpower::phal::exception::DEVTREE_ATTR_READ_FAIL) {
+			log(level::ERROR, "Attribute Type not found for the "
+					  "target at given location code");
+			return std::unexpected(
```

Comment:
Why not like `ATTR_TYPE_NOT_FOUND` like below case.

Reply:
We can return ATTR_TYPE_NOT_FOUND also, but as we have the exact reason i returned that. 
shall i change it to ATTR_TYPE_NOT_FOUND?

Reply:
I think DEVTREE_ATTR_READ_FAIL makes more sense here because we won't be able to determine whether the attribute exists or if it failed to read due to some filesystem issues.